// eventsManager/lib/Channel.js
class Channel {

    /**
     * @type {Map}
     */
    subscribers;

    /**
     * Class constructor
     */
    constructor() {

        // init subscribers
        this.subscribers = new Map();

    }

    /**
     * Class destructor
     */
    destructor() {

        // clear subscribers
        this.subscribers.clear();

    }

    /**
     * Add subscriber
     * @param {Subscriber} subscriber
     */
    addSubscriber(subscriber) {

        // add subscriber
        if (!this.subscribers.has(subscriber)) {
            this.subscribers.set(subscriber, subscriber)
        }

    }

    /**
     * Remove subscriber
     * @param {Subscriber} subscriber
     */
    removeSubscriber(subscriber) {
        this.subscribers.delete(subscriber);
    }

    /**
     * Returns true if at least one subscriber
     * @returns {boolean}
     */
    hasSubscribers() {
        return this.subscribers.size > 0;
    }

    /**
     * Get channel subscribers
     * @returns {Iterator.<Subscriber>}
     */
    getSubscribers() {
        return this.subscribers.values();
    }

}

module.exports = Channel;
